src/economy/shipping_schedule.cc

```diff
@@ -0,0 +1,1704 @@
+              Weighting factors to tweak the algorithm's decisions
+*******************************************************************************/
+
+// Performance tradeoff in the First Pass
+constexpr uint32_t kActualDurationsRecalculationInterval = 60 * 1000;
+
+// Give a ship the highest score for assigning wares to it if it will
+// arrive at the port of interest within the next 10 s,
+// or the lowest score if this will take longer than 10 min
+constexpr Duration kWonderfullyShortDuration = 10 * 1000;   // 10 s
+constexpr Duration kHorriblyLongDuration = 10 * 60 * 1000;  // 10 min
+
+// Only assign wares to a ship in 5.1 if it's score is higher than a certain
+// threshold based on minimal distance. Ships with lower scores will only be
+// accepted if not enough idle ships are found.
+constexpr uint16_t kMinScoreForImmediateAcceptFactor = 14;  // NOCOM needs testing
+
+// Average sailing-time distance for two ports to be considered "close by" in 5.4
+constexpr int16_t kDockGroupMaxDistanceFactor = 12 * 1800;
+
+// Distributing idle ships between ports in the Sixth Pass:
+// Sailing-time distance between a ship and a port so that the ship is considered
+// to be within convenient reach when the port needs a ship
+constexpr int16_t kNearbyDockMaxDistanceFactor = 8 * 1800;
+
+/*******************************************************************************
+                             Actual implementation
+*******************************************************************************/
+
+// #define sslog(...) if (g_verbose) log(__VA_ARGS__);
+#define sslog(...) log("NOCOM: " __VA_ARGS__);  // NOCOM
+
+ShippingSchedule::ShippingSchedule(ShipFleet& f) : fleet_(f), last_updated_(0), loader_(nullptr) {
+	assert(!fleet_.active());
+}
+
+bool ShippingSchedule::empty() const {
+	for (const auto& pair : plans_) {
+		if (!pair.second.empty()) {
+			return false;
+		}
+	}
+	return true;
+}
+
+void ShippingSchedule::ship_arrived(Game& game, Ship& ship, PortDock& port) {
+	sslog(
+	   "\nShippingSchedule::ship_arrived(%s at %u)\n", ship.get_shipname().c_str(), port.serial());
+	auto plan = plans_.find(&ship);
+	assert(plan != plans_.end());
+	const size_t plan_size = plan->second.size();
+	assert(plan_size);
+
+	const SchedulingState& ss = plan->second.front();
+	assert(ss.dock == &port);
+
+	if (ss.expedition) {
+		sslog("Loading expedition\n\n");
+		assert(port.expedition_ready_);
+		assert(ship.get_nritems() == 0);
+		assert(plan_size == 1);  // no planning beyond the expedition
+		assert(ss.load_there.empty());
+
+		std::vector<Worker*> workers;
+		std::vector<WareInstance*> wares;
+		port.expedition_bootstrap_->get_waiting_workers_and_wares(
+		   game, port.owner().tribe(), &workers, &wares);
+		for (Worker* worker : workers) {
+			ship.add_item(game, ShippingItem(*worker));
+		}
+		for (WareInstance* ware : wares) {
+			ship.add_item(game, ShippingItem(*ware));
+		}
```

Comment:
This block should become a ExpeditionBootstrap::load_ship or Ship::boostrap_expedition  Function

Reply:
ExpeditonBootstrap is currently a pure container that stores items and handles the economy-side of requesting stuff. Previously loading expeditions onto ships was handled by the PortDock in the same function that also took care of loading cargo and I kept this design.